state-chain/pallets/cf-ingress-egress/src/lib.rs

```diff
@@ -1082,6 +1122,43 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 			.find(|ccm| ccm.broadcast_id == broadcast_id)
 			.cloned()
 	}
+
+	/// Withholds the fee for a given amount.
+	///
+	/// Returns the remaining amount after the fee has been withheld.
+	fn withhold_transaction_fee(
+		ingress_or_egress: IngressOrEgress,
+		asset: TargetChainAsset<T, I>,
+		available_amount: TargetChainAmount<T, I>,
+	) -> TargetChainAmount<T, I> {
+		let tracked_data = T::ChainTracking::get_tracked_data();
+		let fee_estimate = match ingress_or_egress {
+			IngressOrEgress::Ingress => tracked_data.estimate_ingress_fee(asset),
+			IngressOrEgress::Egress => tracked_data.estimate_egress_fee(asset),
+		};
+
+		let (remaining_amount, fee_estimate) =
+			T::AssetConverter::convert_asset_to_approximate_output(
+				asset,
+				available_amount,
+				<T::TargetChain as Chain>::GAS_ASSET,
+				fee_estimate,
+			)
+			.unwrap_or_else(|| {
+				log::warn!(
+					"Unable to convert input to gas for input of {:?} ${:?}. Ignoring transaction fees.",
+					available_amount,
+					asset,
+				);
+				(available_amount, Zero::zero())
+			});
+
+		WithheldTransactionFees::<T, I>::mutate(<T::TargetChain as Chain>::GAS_ASSET, |fees| {
```

Comment:
we might as well just use StorageValue and not a Map since we only store Gas asset fees her so only one entry in the map.